autodiff/reverse/var/var.hpp

```diff
@@ -1091,42 +1374,78 @@ struct Variable
-template<typename T> bool operator>=(const ExprPtr<T>& l, const Variable<T>& r) { return l->val >= r.expr->val; }
-template<typename T> bool operator<(const ExprPtr<T>& l, const Variable<T>& r) { return l->val < r.expr->val; }
-template<typename T> bool operator>(const ExprPtr<T>& l, const Variable<T>& r) { return l->val > r.expr->val; }
-
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator==(const U& l, const Variable<T>& r) { return l == r.expr; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator!=(const U& l, const Variable<T>& r) { return l != r.expr; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator<=(const U& l, const Variable<T>& r) { return l <= r.expr; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator>=(const U& l, const Variable<T>& r) { return l >= r.expr; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator<(const U& l, const Variable<T>& r) { return l < r.expr; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator>(const U& l, const Variable<T>& r) { return l > r.expr; }
-
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator==(const Variable<T>& l, const U& r) { return l.expr == r; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator!=(const Variable<T>& l, const U& r) { return l.expr != r; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator<=(const Variable<T>& l, const U& r) { return l.expr <= r; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator>=(const Variable<T>& l, const U& r) { return l.expr >= r; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator<(const Variable<T>& l, const U& r) { return l.expr < r; }
-template<typename T, typename U, EnableIf<isArithmetic<U>>...> bool operator>(const Variable<T>& l, const U& r) { return l.expr > r; }
+
+struct ExprValue {
+  template<typename T, EnableIf<isArithmetic<T>>...> static auto value(const T& t) { return t; }
+  template<typename T> static auto value(const ExprPtr<T>& t) { return t->val; }
+  template<typename T> static auto value(const Variable<T>& t) { return t.expr->val; }
+};
+
+template<typename T, typename U>
+using expr_common_t = std::common_type_t<decltype(ExprValue::value(std::declval<T>())), decltype(ExprValue::value(std::declval<U>()))>;
+
+template<class> struct sfinae_true : std::true_type {};
+template<typename T> static auto is_expr_test(int) -> sfinae_true<decltype(ExprValue::value(std::declval<T>()))>;
+template<typename T> static auto is_expr_test(long) -> std::false_type;
+template<typename T> struct is_expr : decltype(is_expr_test<T>(0)) {};
+template<typename T> constexpr bool is_expr_v = is_expr<T>::value;
+
+template<typename T>
+struct CoerceExpr {
```

Comment:
These `coerce` methods can be free (template) functions (instead of static methods of `CoerceExpr`). The name `CoerceExpr` can also confuse, because `SomethingExpr` in `var.hpp` is usually a new type derived from `Expr`.